src/System.Private.ServiceModel/src/System/ServiceModel/UpnEndpointIdentity.cs

```diff
@@ -53,6 +57,110 @@ internal UpnEndpointIdentity(WindowsIdentity windowsIdentity)
             _upnSid = windowsIdentity.User;
             _hasUpnSidBeenComputed = true;
         }
+
+        internal override void EnsureIdentityClaim()
+        {
+            if (_windowsIdentity != null)
+            {
+                lock (_thisLock)
+                {
+                    if (_windowsIdentity != null)
+                    {
+                        Initialize(Claim.CreateUpnClaim(GetUpnFromWindowsIdentity(_windowsIdentity)));
+                        _windowsIdentity.Dispose();
+                        _windowsIdentity = null;
+                    }
+                }
+            }
+        }
+
+        string GetUpnFromWindowsIdentity(WindowsIdentity windowsIdentity)
+        {
+            string downlevelName = null;
+            string upnName = null;
+ 
+            try
+            {
+                downlevelName = windowsIdentity.Name;
+ 
+                if (IsMachineJoinedToDomain())
```

Comment:
I see it will always throw *and catch* PNSE and then fallback to the downlevelName.  I know the number of calls to this could be low, but it's annoying to have unconditional exceptions being thrown.  Can that call to IsMachineJoinedToDomain itself be selectively commented out?  It does feel weird to differ from desktop unconditionally and silently, but maybe we have no other options.  Do we know which real-life scenarios will fail and which might still pass (falling back to downlevelName)?

---
In reply to: [77481060](https://github.com/dotnet/wcf/pull/1502#discussion_r77481060) [](ancestors = 77481060)